drivers/led_strip/ws2812_rpi_pico_pio.c

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (c) 2023 TOKITA Hiroshi
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/led_strip.h>
+#include <zephyr/drivers/misc/pio_rpi_pico/pio_rpi_pico.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/dt-bindings/led/led.h>
+#include <zephyr/sys_clock.h>
+
+#define DT_DRV_COMPAT worldsemi_ws2812_rpi_pico_pio
+
+#define LOG_LEVEL CONFIG_LED_STRIP_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ws2812_rpi_pico_pio);
+
+#define CYCLES_PER_BIT 10
+
+struct ws2812_rpi_pico_pio_config {
+	const struct device *piodev;
+	const struct pinctrl_dev_config *pcfg;
+	const uint32_t pin;
+	const uint8_t num_colors;
+	const uint32_t frequency;
+	const uint8_t *color_mapping;
+};
+
+RPI_PICO_PIO_DEFINE_PROGRAM(ws2812, 0, 3,
+		/*  .wrap_target                          */
+	0x6221, /*  0: out    x, 1            side 0 [2]  */
+	0x1123, /*  1: jmp    !x, 3           side 1 [1]  */
+	0x1400, /*  2: jmp    0               side 1 [4]  */
+	0xa442, /*  3: nop                    side 0 [4]  */
+		/*  .wrap                                 */
+);
+
+static void ws2812_program_init(PIO pio, uint32_t sm, uint32_t pin, float freq)
+{
+	float clkdiv = sys_clock_hw_cycles_per_sec() / (CYCLES_PER_BIT * freq);
+	uint32_t offset;
+
+	offset = pio_add_program(pio, &ws2812_program);
+
+	pio_sm_config sm_config = pio_get_default_sm_config();
+
+	sm_config_set_wrap(&sm_config, offset + RPI_PICO_PIO_GET_WRAP_TARGET(ws2812),
+			   offset + RPI_PICO_PIO_GET_WRAP(ws2812));
+	sm_config_set_sideset(&sm_config, 1, false, false);
+	sm_config_set_sideset_pins(&sm_config, pin);
+	sm_config_set_out_shift(&sm_config, false, true, 24);
+	sm_config_set_fifo_join(&sm_config, PIO_FIFO_JOIN_TX);
+	sm_config_set_clkdiv(&sm_config, clkdiv);
+	pio_sm_set_consecutive_pindirs(pio, sm, pin, 1, true);
+	pio_sm_init(pio, sm, offset, &sm_config);
+	pio_sm_set_enabled(pio, sm, true);
+}
+
+static int ws2812_rpi_pico_pio_update_rgb(const struct device *dev, struct led_rgb *pixels,
+					  size_t num_pixels)
+{
+	const struct ws2812_rpi_pico_pio_config *config = dev->config;
+	PIO pio = pio_rpi_pico_get_pio(config->piodev);
+
+	for (size_t i = 0; i < num_pixels; i++) {
+		uint32_t color = 0;
+
+		for (size_t j = 0; j < config->num_colors; j++) {
+			switch (config->color_mapping[j]) {
+			/* White channel is not supported by LED strip API. */
+			case LED_COLOR_ID_WHITE:
+				color |= 0;
+				break;
+			case LED_COLOR_ID_RED:
+				color |= pixels[i].r << (8 * (2 - j));
+				break;
+			case LED_COLOR_ID_GREEN:
+				color |= pixels[i].g << (8 * (2 - j));
+				break;
+			case LED_COLOR_ID_BLUE:
+				color |= pixels[i].b << (8 * (2 - j));
+				break;
```

Comment:
The `j` is from index in color_mapping array. It takes from 0, 1 or 2.
So, if color_mapping is [RED, GREEN, BLUE], 
the calculation is (pixels[i].r << (8 * (2)) | pixels[i].g << (8 * (1) | pixels[i].b << (8 * (0)).
Otherwise,  color_mapping is [GREEN, RED, BLUE] case,
the calculation is (pixels[i].r << (8 * (1)) | pixels[i].g << (8 * (2) | pixels[i].b << (8 * (0)).

It sort the bits correctly based on the color_mapping settings.